stan/math/prim/fun/complex_schur_decompose.hpp

```diff
@@ -0,0 +1,64 @@
+#ifndef STAN_MATH_PRIM_FUN_COMPLEX_SCHUR_DECOMPOSE_HPP
+#define STAN_MATH_PRIM_FUN_COMPLEX_SCHUR_DECOMPOSE_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return the unitary matrix of the complex Schur decomposition of the
+ * specified square matrix.
+ *
+ * The complex Schur decomposition of a square matrix `A` produces a
+ * complex unitary matrix `U` and a complex upper-triangular Schur
+ * form matrix `T` such that `A = U * T * inv(U)`.  Further, the
+ * unitary matrix's inverse is equal to its conjugate transpose,
+ * `inv(U) = U*`, where `U*(i, j) = conj(U(j, i))`
+ *
+ * @tparam value type of matrix
+ * @param m real matrix to decompose
+ * @return complex unitary matrix of the complex Schur decomposition of the
+ * specified matrix
+ * @see complex_schur_decompose_t
+ */
+template <typename M, require_eigen_dense_dynamic_t<M>* = nullptr>
+inline Eigen::Matrix<complex_return_t<scalar_type_t<M>>, -1, -1>
+complex_schur_decompose_u(const M& m) {
+  check_nonzero_size("complex_schur_decompose_u", "m", m);
```

Comment:
Shouldn't size 0 just return a 0 x 0 matrix for both components?  If `A`, `U` and `T` are size 0 x 0, then `A = U * T * inv(U)` is also 0 x 0.  The inverse of a 0 x 0 matrix is just the unique 0 x 0 matrix.

Reply:
This is an implementation choice I don't have a strong opinion on, but I should note that `scipy` throws an error on size-zero inputs.